In [9]:
from random import random

def funcion(x, y):
    sum1 = x**2 * (4 - 2.1*x**2 + x**4/3.0)
    sum2 = x*y
    sum3 = y**2 * (-4 + 4*y**2)
    return sum1 + sum2 + sum3

def aleatorio(inf, sup):
    return random()*(sup-inf) + inf

class Particula:
    inercia = 1.4
    cognitiva = 2.0
    social = 2.0
    infx = -10.0
    supx = 10.0
    infy = -10.0
    supy = 10.0
    ajusteV = 100.0

    def __init__(self):
        self.x = aleatorio(Particula.infx, Particula.supx)
        self.y = aleatorio(Particula.infy, Particula.supy)
        self.vx = aleatorio(Particula.infx/Particula.ajusteV, Particula.supx/Particula.ajusteV)
        self.vy = aleatorio(Particula.infy/Particula.ajusteV, Particula.supy/Particula.ajusteV)
        self.xLoc = self.x
        self.yLoc = self.y
        self.valorLoc = funcion(self.x, self.y)

    def actualizaVelocidad(self, xGlob, yGlob):
        cogX = Particula.cognitiva*random()*(self.xLoc - self.x)
        socX = Particula.social*random()*(xGlob - self.x)
        self.vx = Particula.inercia*self.vx + cogX + socX
        cogY = Particula.cognitiva*random()*(self.yLoc - self.y)
        socY = Particula.social*random()*(yGlob - self.y)
        self.vy = Particula.inercia*self.vy + cogY + socY

    def actualizaPosicion(self):
        self.x = self.x + self.vx
        self.y = self.y + self.vy
        self.x = max(self.x, Particula.infx)
        self.x = min(self.x, Particula.supx)
        self.y = max(self.y, Particula.infy)
        self.y = min(self.y, Particula.supy)
        valor = funcion(self.x, self.y)
        if valor < self.valorLoc:
            self.xLoc = self.x
            self.yLoc = self.y
            self.valorLoc = valor

def enjambreParticulas(particulas, iteraciones, reduccionInercia):
    mejorParticula = min(particulas, key=lambda p: p.valorLoc)
    xGlob = mejorParticula.xLoc
    yGlob = mejorParticula.yLoc
    valorGlob = mejorParticula.valorLoc

    for iter in range(iteraciones):
        for p in particulas:
            p.actualizaVelocidad(xGlob, yGlob)
            p.actualizaPosicion()

        mejorParticula = min(particulas, key=lambda p: p.valorLoc)
        if mejorParticula.valorLoc < valorGlob:
            xGlob = mejorParticula.xLoc
            yGlob = mejorParticula.yLoc
            valorGlob = mejorParticula.valorLoc

        Particula.inercia *= reduccionInercia

    return (xGlob, yGlob, valorGlob)

nParticulas = 10
iteraciones = 100
redInercia = 0.9

particulas = [Particula() for i in range(nParticulas)]

print(enjambreParticulas(particulas, iteraciones, redInercia))

(-0.08984201449225639, 0.7126564027604881, -1.0316284534898774)


El código PSO encontró un resultado más cercano al valor real (-1.031628 vs -1.031570) y lo hizo evaluando muchos menos puntos. La búsqueda estructurada recorre toda la grilla de forma sistemática pero queda limitada por el tamaño del paso, mientras que el PSO busca de forma inteligente y llega a una mejor solución con menos cálculos.

minimos globales:

min{x^2 (4 - 2.1 x^2 + x^4/3) + x y + y^2 (-4 + 4 y^2)}≈-1.03163 en (x, y)≈(-0.089842, 0.712656)


min{x^2 (4 - 2.1 x^2 + x^4/3) + x y + y^2 (-4 + 4 y^2)}≈-1.03163 en (x, y)≈(0.089842, -0.712656)

In [ ]:
def funcion(x, y):
    return x**2 * (4 - 2.1 * x**2 + x**4 / 3.0) + x * y + y**2 * (-4 + 4 * y**2)


def busqueda_estructurada(Xmin, Xmax, Ymin, Ymax, step):
    mejor_x = Xmin
    mejor_y = Ymin
    mejor_valor = funcion(mejor_x, mejor_y)

    x = Xmin
    while x <= Xmax:
        y = Ymin
        while y <= Ymax:
            valor = funcion(x, y)

            if valor < mejor_valor:
                mejor_valor = valor
                mejor_x = x
                mejor_y = y

            y += step
        x += step

    return mejor_x, mejor_y, mejor_valor


Xmin = -10.0
Xmax = 10.0
Ymin = -10.0
Ymax = 10.0
step = 0.01

resultado = busqueda_estructurada(Xmin, Xmax, Ymin, Ymax, step)

print("Mejor solución encontrada con programación estructurada:")
print("el resultado de X:", resultado[0])
print("el resultado de Y:", resultado[1])
print("f(x,y) = ", resultado[2])

Mejor solución encontrada con programación estructurada:
x = 0.0899999999998311
y = -0.7100000000001694
f(x,y) = -1.031570363853008
